zkevm-circuits/src/bin/stats/halo2_stats.rs

```diff
@@ -0,0 +1,374 @@
+        // The estimated peak memory formula comes from point 9 of the analysis, which is the step
+        // of proving that needs the most memory (after that step, allocations start getting freed)
+
+        // number of "elliptic curve points"
+        let m_c = 2;
+        // number of "field elements"
+        let m_s = 1 + 2 * c_f + 2 * c_p + 2 * c_i + c_a + 3 * c_l + c_pg;
+        // number of "field elements" that will be extended by "* e"
+        let m_e = 4 + c_f + c_p + c_pg + c_i + c_a + 3 * (c_l > 0) as usize;
+        let unit = 2 * m_c + m_s + e * m_e;
+        unit * 2usize.pow(k) * field_bytes
+    }
+}
+
+// Return the stats in `meta`, accounting only for the circuit delta from the last aggregated stats
+// in `agg`.
+// Adaptaed from Scroll https://github.com/scroll-tech/zkevm-circuits/blob/7d9bc181953cfc6e7baf82ff0ce651281fd70a8a/zkevm-circuits/src/util.rs#L294
+pub(crate) fn circuit_stats<F: Field>(
+    agg: &CircuitStats,
+    meta: &ConstraintSystem<F>,
+) -> CircuitStats {
+    let max_phase = meta
+        .advice_column_phase()
+        .iter()
+        .skip(agg.num_advice_columns)
+        .max()
+        .copied()
+        .unwrap_or_default();
+
+    let rotations = meta
+        .advice_queries()
+        .iter()
+        .skip(agg.num_advice_queries)
+        .map(|(_, q)| q.0)
+        .collect::<BTreeSet<i32>>();
+
+    let num_fixed_columns = meta.num_fixed_columns() - agg.num_fixed_columns;
+    let num_lookups = meta.lookups().len() - agg.num_lookups;
+    let num_shuffles = meta.shuffles().len() - agg.num_shuffles;
+    let num_advice_columns = meta.num_advice_columns() - agg.num_advice_columns;
+    let num_instance_columns = meta.num_instance_columns() - agg.num_instance_columns;
+    let num_selectors = meta.num_selectors() - agg.num_selectors;
+    let num_permutation_columns =
+        meta.permutation().get_columns().len() - agg.num_permutation_columns;
+
+    CircuitStats {
+        num_constraints: meta
+            .gates()
+            .iter()
+            .skip(agg.num_gates)
+            .map(|g| g.polynomials().len())
+            .sum::<usize>(),
+        num_fixed_columns,
+        num_lookups,
+        num_shuffles,
+        num_advice_columns,
+        num_instance_columns,
+        num_selectors,
+        num_permutation_columns,
+        degree: meta.degree(),
+        blinding_factors: meta.blinding_factors(),
+        num_challenges: meta.num_challenges() - agg.num_challenges,
+        max_phase,
+        num_rotation: rotations.len(),
+        min_rotation: rotations.first().cloned().unwrap_or_default(),
+        max_rotation: rotations.last().cloned().unwrap_or_default(),
+        num_verification_ecmul: num_advice_columns
+            + num_instance_columns
+            + num_permutation_columns
+            + num_shuffles
+            + num_selectors
+            + num_fixed_columns
+            + 3 * num_lookups
+            + rotations.len(),
```

Comment:
I think this part is not accurate:
- `num_instance_columns` will not contribute when we are using KZG commitment scheme
- grand products in permutaton argument are not counted in (`div_ceil(num_permutation_columns, degree - 2)`)
- quotients are not counted in (`degree - 1`)
- `num_selectors` might be ambiguous cause it might be compressed
- batch open scheme matters (GWC as `rotations.len()`, SHPLONK as `2`).

Reply:
> `num_selectors` might be ambiguous cause it might be compressed

Yes, for this bullet I'm talking about the link you share.

Reply:
So we might have extra fixed columns, but I guess some imprecision in the nbs or just having order of magnitudes may still be fine

Reply:
yeah make sense, Edu also added [the comment](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1763/commits/d9fc61b7bff34c6c7b11905f309c1d18488cd42d#diff-0e9ee29187a7ec01aa350218b105d413b8f63ff6ac04747eafe63003a4a6ec6bR299) so we know this is an estimated upper bound